src/msgs/measurements.rs

```diff
@@ -281,49 +413,470 @@ bitflags! {
+        ParseMeasurementBlockError::Measurement(e)
+    }
+}
+
+impl MeasurementBlock {
+    fn serialized_size(&self) -> usize {
+        4 + self.measurement.serialized_size()
+    }
+
+    fn write(&self, w: &mut Writer) -> Result<usize, BufferFullError> {
+        w.put(self.index)?;
+        w.put(self.measurement_spec_selected.bits())?;
+        w.put_u16(self.measurement.serialized_size().try_into().unwrap())?;
+        self.measurement.write(w)
+    }
+
+    fn read(
+        r: &mut Reader,
+    ) -> Result<MeasurementBlock, ParseMeasurementBlockError> {
+        let index = r.get_byte()?;
+        let spec = r.get_byte()?;
+        if spec != MeasurementSpec::DMTF.bits() {
+            return Err(ParseMeasurementBlockError::InvalidMeasurementSpec);
+        }
+        // It's actually unnecessary to read the size, since the Measurement is
+        // self describing. We use this to validate the size of the read Measurement
+        // matches what was put on the wire.
+        let size = r.get_u16()? as usize;
+        let offset = r.byte_offset();
+        let measurement = DmtfMeasurement::read(r)?;
+
+        if size != r.byte_offset() - offset {
+            return Err(ParseMeasurementBlockError::MeasurementSizeMismatch);
+        }
+
+        Ok(MeasurementBlock {
+            index,
+            measurement_spec_selected: MeasurementSpec::DMTF,
+            measurement,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum WriteMeasurementsError {
+    BufferFull,
+    WriteOpaqueElement(WriteOpaqueElementError),
+}
+
+impl From<BufferFullError> for WriteMeasurementsError {
+    fn from(_: BufferFullError) -> Self {
+        WriteMeasurementsError::BufferFull
+    }
+}
+
+impl From<WriteOpaqueElementError> for WriteMeasurementsError {
+    fn from(e: WriteOpaqueElementError) -> Self {
+        WriteMeasurementsError::WriteOpaqueElement(e)
+    }
+}
+
+#[derive(Debug)]
+pub enum ParseMeasurementBlocksError {
+    MaxSizeExceeded,
+    MeasurementSizeMismatch,
+    Block(ParseMeasurementBlockError),
+    Read(ReadError),
+}
+
+impl From<ParseMeasurementBlockError> for ParseMeasurementBlocksError {
+    fn from(e: ParseMeasurementBlockError) -> Self {
+        ParseMeasurementBlocksError::Block(e)
+    }
+}
+
+impl From<ReadError> for ParseMeasurementBlocksError {
+    fn from(e: ReadError) -> Self {
+        ParseMeasurementBlocksError::Read(e)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct MeasurementBlocks {
+    pub len: usize,
+    pub blocks: [Option<MeasurementBlock>; config::MAX_MEASUREMENT_BLOCKS],
+}
+
+impl Default for MeasurementBlocks {
+    fn default() -> Self {
+        const BLOCK: Option<MeasurementBlock> = None;
+        MeasurementBlocks {
+            len: 0,
+            blocks: [BLOCK; config::MAX_MEASUREMENT_BLOCKS],
+        }
+    }
+}
+
+impl MeasurementBlocks {
+    // All options returned are `Some` variants.
+    //
+    // TODO: Use MaybeUninit and return a `&[MeasurementBlock]`?
```

Comment:
I had the same thought as I was reading this. It would require making `len` and `blocks` private - would that be onerous? How does a caller construct a `MeasurementBlocks` today (if they want to create one to write)?

Reply:
The only user right now is [tests](https://github.com/oxidecomputer/spdm/pull/41/files#diff-d2e83961de0a47da6f0616ad312695b32e0aebf44b67c679843a5db45e7fa4f8R1081-R1095). We could certainly add a `push` operation. I'll leave this open for a follow up though.